plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+
+    private final MetadataBuilder metadataBuilder;
+
+    private final MetadataUpdater metadataUpdater;
+
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    public Worker(final Repository repository,
+                  final boolean update,
+                  final String sql,
+                  final Map<String, Object> sqlParams)
+    {
+      this.update = update;
+      this.sql = sql;
+      this.sqlParams = sqlParams;
+      this.storageFacet = repository.facet(StorageFacet.class);
+      this.mavenFacet = repository.facet(MavenFacet.class);
+      this.mavenPathParser = mavenFacet.getMavenPathParser();
+      this.metadataBuilder = new MetadataBuilder();
+      this.metadataUpdater = new MetadataUpdater(mavenFacet);
+      this.documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    public void rebuild(final StorageTx tx)
+    {
+      final ODatabaseDocumentTx readDatabase = tx.getDb();
+      readDatabase.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  try {
+                    final ODocument doc = (ODocument) iRecord;
+                    final String groupId = doc.field("groupId", OType.STRING);
+                    final String artifactId = doc.field("artifactId", OType.STRING);
+                    final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                    final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                    if (groupChange) {
+                      if (currentGroupId != null) {
+                        processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                      }
+                      currentGroupId = groupId;
+                      metadataBuilder.onEnterGroupId(groupId);
+                    }
+                    rebuildAbV((ORID) sqlParams.get("bucket"), groupId, artifactId, baseVersions);
+                    return true;
+                  }
+                  finally {
+                    ODatabaseRecordThreadLocal.INSTANCE.set(readDatabase); // always reset thread DB
+                  }
+                }
+
+                @Override
+                public void end() {
+                  if (currentGroupId != null) {
+                    processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                  }
+                }
+              }
+          )
+      ).execute(sqlParams);
+    }
+
+    /**
+     * Method processing in separate TX/DB, performs writes (and mangles ThreadLocal DB, so caller should restore it).
+     * Handles A and bV on metadataBuilder.
+     */
+    private void rebuildAbV(final ORID bucket,
```

Comment:
Can we have a more explicit name please? I know what A and bV are here, but no one looking at this two weeks from now would know without actually having to read the code ;)